src/engine_render.c

```diff
@@ -8002,12 +8010,21 @@ void draw_keepsprite_unscaled_in_buffer(unsigned short kspr_n, short angle, unsi
         {
             return;
         }
-        kspr = &kspr_arr[a3 + quarter * kspr_arr->FramesCount];
+        kspr = &kspr_arr[field48 + quarter * kspr_arr->FramesCount];
         fill_w = kspr->SWidth;
         fill_h = kspr->SHeight;
-        keepsprite_id = a3 + quarter * kspr->FramesCount + kspr_idx;
-        if (keepsprite_id >= KEEPSPRITE_LENGTH) {
-            return; // don't try to draw shadows for custom sprites or game will crash
+        keepsprite_id = field48 + quarter * kspr->FramesCount + kspr_idx;
+        if (keepsprite_id >= KEEPERSPRITE_ADD_OFFSET)
+        {
+            sprite_data = keepersprite_add[keepsprite_id - KEEPERSPRITE_ADD_OFFSET];
+        }
+        else if (keepsprite_id >= KEEPSPRITE_LENGTH)
+        {
+            return; // WTF?!!
```

Comment:
@SimTheVoid This was implemented as a workaround in #1700 , otherwise it would crash. This is code that could be removed when the shadows are implemented.

Reply:
It is to protect from crash on invalid values (less than custom sprites, but more than originals)

Reply:
So instead of 'wtf' we could go with something like this?
```c
ERRORLOG("Sprite %d outside of valid range.",keepsprite_id);
```